File: Deep-Learning-for-Beginners-master/R/ch.5/Dropout.R
Dropout <- function(y, ratio) {
  d <- dim(y)
  m <- d[1]
  n <- d[2]
  ym <- array(0, d)

  num <- round(m * n * (1 - ratio))
  idx <- sample(m * n, num)
  ym[idx] <- 1 / (1 - ratio)
  return(ym)
}

File: Deep-Learning-for-Beginners-master/Go/utils/dropout.go
package utils

import (
	"gonum.org/v1/gonum/mat"
	"math"
	"math/rand"
)

func Dropout(recv *mat.VecDense, ratio float64) {
	size := recv.Len()
	numSurvived := int(math.Round(float64(size) * (1.0 - ratio)))
	indexes := rand.Perm(size)
	zeros := mat.NewVecDense(size, nil)
	for i := 0; i < numSurvived; i++ {
		zeros.SetVec(indexes[i], 1.0 / (1.0 - ratio))
	}
	recv.MulElemVec(recv, zeros)
}
